### ISSUE TITLE
[Feature] Containers should have access to the UID and GID of the user who launched them

### ISSUE THREAD

**Issue by marcosdiez** [docker/docker#20812]
I believe this simple feature, a user on every container which inherits the UID and GID of the user who launched the container, would make things very flexible.

This would allow a new breed of containers: containers that can only read and write files the user who launched them would have access in first place.

It seems very useful for compilers, text editors or any other end user app in a container.

These kind of containers would be safe enough to be run by any non root user on a system.

Please consider implementing this feature!

**Comment by thaJeztah** [comment]
I don't think we can support this; there's various things to talk into consideration here. One of those is that containers are started on the host where the *daemon* runs, which may be a different server altogether (you can have the client locally, and the daemon running, for example, on a server in the cloud).

Also, *which* files should be accessible by the container? Processes in the container run in an isolated environment *deliberately*, so that they don't have access to anything, unless you give them access.

However, you can do something like;

```bash

docker run --user=my-userid:my-group-id -v $HOME:/home/myusername -it ubuntu bash
```

Which will start a container using the same uid/gid as you, and mount your home-directory inside the container.

I'm closing this, for the reasons above, but feel free to comment here after I've closed

**Comment by marcosdiez** [comment]
Hi !
I think I was not clear enough.
First of all, I completely agree it's an issue if the user and the deamon are in different machines. Let's ignore that for a while. 

Second, of course the container should only have access to what is exported with --volume.

The idea would be something like:

docker run -v .:/data_output compile_the_app

This would allow for example apps that need root to be compiled (which is wrong in first place, but some npm libraries are that unfortunately that awful) but make sure the output would be stored on the host's filesystem with the permissions of the user who launched the container.

In other words this would wrap this misbehaving app into a container and one would get the files these apps generate out of the container, except that it would work with any user.

I currently have solved this problem like this:

if [ -n "${TARGET_UID}" ] && [ -n "${TARGET_GID}" ]
then
    echo "Setting ownership of the build to ${TARGET_UID}:${TARGET_GID}"
    chown -R ${TARGET_UID}:${TARGET_GID} /srv/build
fi

but that is not bullet proof for it fails if the build (which runs inside of the container) crashes.

Don't you think more users would want such functionality ? (despite the remote daemon issue)

**Comment by thaJeztah** [comment]
Automatically changing ownership of files on the host can be very dangerous, for example, consider running `docker run -v $(pwd):/home ...` and all of a sudden you're locked out of your home-directory, because docker changed the owner of your home-directory to `root:root`.

We had a proposal for this; https://github.com/docker/docker/pull/14632, continued in https://github.com/docker/docker/pull/16767. Which was rejected because it could lead to problematic issues; https://github.com/docker/docker/pull/16767#issuecomment-157937286

You may be interested in the new User Namespace feature in Docker 1.10, which allows you to map a non-privileged uid/gid on the host to "root" inside the container; https://docs.docker.com/engine/reference/commandline/daemon/#daemon-user-namespace-options

**Comment by gtirloni** [comment]
While the user namespace feature is a welcome improvement, it is too complicated for end users at this time. 

End users need a single command line option to `docker run` that ensures the data written to the shared volume from inside the container is owned by their users, without having to creating custom entrypoint scripts or fiddle with dockerd options, systemd, etc.

I understand the technical points for why the proposed solutions won't work today (unless, as @cpuguy83 mentioned a few times, we get a new FS with proper support for what is being requested here), but just wanted to emphasize the UX aspect of this.

Would OverlayFS 2.x be a good target for this functionality?

**Comment by cpuguy83** [comment]
@gtirloni See "ShiftFS" from James Bottomley.
This has been floating around amongs kernel maintainers for most of 2016 and hopefully will see some light in the kernel next year.

**Comment by gtirloni** [comment]
That seems promising, thank you.
